File: data_structures/stack.js
/**
 * Created by dmorales on 12/01/2016.
 */


Stack = function() {
    var stack = {}, list = {};

    stack.push = function(elem) {

    };

    stack.add = function(elem) {

    };

    stack.pop = function() {

    };

    stack.peek = function() {
        return list.first();
    };

    stack.size = function() {
        return list.size();
    };

    stack.contains = function (elem, equalsFunction) {

    };

    stack.isEmpty = function () {
        return list.isEmpty();
    };

    stack.clear = function () {
        list.clear();
    };

    stack.forEach = function(callback) {
        list.forEach(callBack);
    };

    stack.toArray = function() {
        return list.toArray();
    };

    stack.equals = function(other, equalsFunction) {

    };

    return stack;

};

var myStack = new Stack();
myStack.add('a');
myStack.add('b');

File: java_scratchpad/Stack.java
package java_scratchpad;

/**
 * Created by darwinmorales on 26/07/2016.
 */
public class Stack<E> {
    private int top = -1;
    private E[] data;


    public Stack(int size) {
        data = (E[]) new Object[size];
    }

    public void push(E e) {
        if (top == data.length - 1) {
            return;
        }
        data[++top] = e;
    }

    public E pop() {
        if (isEmpty()) {
            return null;
        }
        E temp = data[top];
        data[top] = null;
        top--;
        return temp;
    }

    public boolean isEmpty() {
        return top < 0;
    }


    public static void main(String[] args) {
        Stack<String> stack = new Stack<>(3);

        System.out.println( stack.isEmpty());

        stack.push("A");
        stack.push("B");
        stack.push("C");
        stack.push("D");

        System.out.println(stack.pop());
        System.out.println(stack.pop());
        System.out.println(stack.pop());

        System.out.println( stack.isEmpty());


    }

}
